Warning due to ignored return value of std::async

Following build log occurs in CI:

```
[ 37%] Performing build step for 'koinos_mq-Release'
Scanning dependencies of target koinos_mq
[ 16%] Building CXX object libraries/mq/CMakeFiles/koinos_mq.dir/client.cpp.o
[ 33%] Building CXX object libraries/mq/CMakeFiles/koinos_mq.dir/request_handler.cpp.o
[ 50%] Building CXX object libraries/mq/CMakeFiles/koinos_mq.dir/message_broker.cpp.o
/home/travis/.hunter/_Base/69371a6/5a48d05/c107877/Build/koinos_mq/Source/libraries/mq/client.cpp:282:7: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
      std::async(
      ^~~~~~~~~~
1 warning generated.

[ 66%] Linking CXX static library libkoinos_mq.a
```

The relevant `std::async()` call is [here](https://github.com/koinos/koinos-mq-cpp/blob/cfd7544d4603feb539a2f35ce5d37b39acff77aa/libraries/mq/client.cpp#L282-L299).

--- a/libraries/mq/client.cpp
+++ b/libraries/mq/client.cpp
@@ -6,6 +6,7 @@
 #include <map>
 #include <mutex>
 #include <random>
+#include <thread>
 
 namespace koinos::mq {
 
@@ -29,6 +30,7 @@ private:
    error_code prepare();
    void consumer( std::shared_ptr< message_broker > broker );
    std::string random_alphanumeric( std::size_t len );
+   void timeout( std::shared_future< std::string > future, std::string correlation_id, uint64_t timeout_ms );
 
    std::map< std::string, std::promise< std::string > > _promise_map;
    std::mutex                                           _promise_map_mutex;
@@ -223,6 +225,21 @@ void client_impl::consumer( std::shared_ptr< message_broker > broker )
    }
 }
 
+void client_impl::timeout( std::shared_future< std::string > future, std::string correlation_id, uint64_t timeout_ms )
+{
+   auto status = future.wait_for( std::chrono::milliseconds( timeout_ms ) );
+   if ( status != std::future_status::ready )
+   {
+      std::lock_guard< std::mutex > guard( _promise_map_mutex );
+      auto itr = _promise_map.find( correlation_id );
+      if ( itr != _promise_map.end() )
+      {
+         itr->second.set_exception( std::make_exception_ptr( timeout_error( "Request timeout: " + correlation_id ) ) );
+         _promise_map.erase( itr );
+      }
+   }
+}
+
 std::shared_future< std::string > client_impl::rpc( const std::string& service, const std::string& payload, const std::string& content_type, int64_t timeout_ms )
 {
    KOINOS_ASSERT( _running, client_not_running, "Client is not running" );
@@ -235,6 +252,14 @@ std::shared_future< std::string > client_impl::rpc( const std::string& service, 
    msg.data = payload;
    msg.reply_to = _queue_name;
    msg.correlation_id = random_alphanumeric( 32 );
+
+   LOG(debug) << "Sending RPC";
+   LOG(debug) << " -> correlation_id: " << *msg.correlation_id;
+   LOG(debug) << " -> exchange: " << msg.exchange;
+   LOG(debug) << " -> routing_key: " << msg.routing_key;
+   LOG(debug) << " -> content_type: " << msg.content_type;
+   LOG(debug) << " -> reply_to: " << *msg.reply_to;
+   LOG(debug) << " -> data: " << msg.data;
 
    auto err = _writer_broker->publish( msg );
    if ( err != error_code::success )
@@ -260,26 +285,8 @@ std::shared_future< std::string > client_impl::rpc( const std::string& service, 
 
    if ( timeout_ms > 0 )
    {
-      std::async(
-         std::launch::async,
-         [&]( std::shared_future< std::string > future )
-         {
-            auto status = future.wait_for( std::chrono::milliseconds( timeout_ms ) );
-            if ( status != std::future_status::ready )
-            {
-               std::lock_guard< std::mutex > guard( _promise_map_mutex );
-               auto itr = _promise_map.find( *msg.correlation_id );
-               if ( itr != _promise_map.end() )
-               {
-                  itr->second.set_exception( std::make_exception_ptr( timeout_error( "Request timeout" ) ) );
-                  _promise_map.erase( itr );
-               }
-            }
-         },
-         future_val
-      );
+      std::thread( &client_impl::timeout, this, future_val, *msg.correlation_id, timeout_ms ).detach();
    }
-
 
    return future_val;
 }

--- a/libraries/mq/client.cpp
+++ b/libraries/mq/client.cpp
@@ -138,7 +138,7 @@ error_code client_impl::prepare()
 
    if ( ec != error_code::success )
    {
-      LOG(error) << "error while declaring broadcast exchange";
+      LOG(error) << "Error while declaring broadcast exchange";
       return ec;
    }
 
@@ -153,7 +153,7 @@ error_code client_impl::prepare()
 
    if ( ec != error_code::success )
    {
-      LOG(error) << "error while declaring rpc exchange";
+      LOG(error) << "Error while declaring rpc exchange";
       return ec;
    }
 
@@ -167,7 +167,7 @@ error_code client_impl::prepare()
 
    if ( queue_res.first != error_code::success )
    {
-      LOG(error) << "error while declaring temporary queue";
+      LOG(error) << "Error while declaring temporary queue";
       return queue_res.first;
    }
 
@@ -176,7 +176,7 @@ error_code client_impl::prepare()
    ec = _reader_broker->bind_queue( queue_res.second, exchange::rpc, queue_res.second );
    if ( ec != error_code::success )
    {
-      LOG(error) << "error while binding temporary queue";
+      LOG(error) << "Error while binding temporary queue";
       return ec;
    }
 
@@ -196,20 +196,20 @@ void client_impl::consumer( std::shared_ptr< message_broker > broker )
 
       if ( result.first != error_code::success )
       {
-         LOG(error) << "failed to consume message";
+         LOG(error) << "Failed to consume message";
          continue;
       }
 
       if ( !result.second )
       {
-         LOG(error) << "consumption succeeded but resulted in an empty message";
+         LOG(error) << "Consumption succeeded but resulted in an empty message";
          continue;
       }
 
       auto& msg = result.second;
       if ( !msg->correlation_id.has_value() )
       {
-         LOG(error) << "received message without a correlation id";
+         LOG(error) << "Received message without a correlation id";
          continue;
       }
 
@@ -255,11 +255,11 @@ std::shared_future< std::string > client_impl::rpc( const std::string& service, 
 
    LOG(debug) << "Sending RPC";
    LOG(debug) << " -> correlation_id: " << *msg.correlation_id;
-   LOG(debug) << " -> exchange: " << msg.exchange;
-   LOG(debug) << " -> routing_key: " << msg.routing_key;
-   LOG(debug) << " -> content_type: " << msg.content_type;
-   LOG(debug) << " -> reply_to: " << *msg.reply_to;
-   LOG(debug) << " -> data: " << msg.data;
+   LOG(debug) << " -> exchange:       " << msg.exchange;
+   LOG(debug) << " -> routing_key:    " << msg.routing_key;
+   LOG(debug) << " -> content_type:   " << msg.content_type;
+   LOG(debug) << " -> reply_to:       " << *msg.reply_to;
+   LOG(debug) << " -> data:           " << msg.data;
 
    auto err = _writer_broker->publish( msg );
    if ( err != error_code::success )

--- a/libraries/mq/message_broker.cpp
+++ b/libraries/mq/message_broker.cpp
@@ -367,7 +367,7 @@ std::optional< std::string > message_broker_impl::error_info( amqp_rpc_reply_t r
 {
    if ( r.reply_type == AMQP_RESPONSE_NONE )
    {
-      return "missing rpc reply type";
+      return "Missing RPC reply type";
    }
    else if ( r.reply_type == AMQP_RESPONSE_LIBRARY_EXCEPTION )
    {
@@ -385,7 +385,7 @@ std::optional< std::string > message_broker_impl::error_info( amqp_rpc_reply_t r
             snprintf(
                buf,
                bufsize,
-               "server connection error %u, message: %.*s",
+               "Server connection error %u, message: %.*s",
                m->reply_code,
                (int)m->reply_text.len,
                (char*)m->reply_text.bytes
@@ -398,7 +398,7 @@ std::optional< std::string > message_broker_impl::error_info( amqp_rpc_reply_t r
             snprintf(
                buf,
                bufsize,
-               "server channel error %u, message: %.*s",
+               "Server channel error %u, message: %.*s",
                m->reply_code,
                (int)m->reply_text.len,
                (char*)m->reply_text.bytes
@@ -406,7 +406,7 @@ std::optional< std::string > message_broker_impl::error_info( amqp_rpc_reply_t r
             return buf;
          }
          default:
-            snprintf( buf, bufsize, "unknown server error, method id 0x%08X", r.reply.id );
+            snprintf( buf, bufsize, "Unknown server error, method ID 0x%08X", r.reply.id );
             return buf;
       }
    }

--- a/libraries/mq/message_broker.cpp
+++ b/libraries/mq/message_broker.cpp
@@ -175,7 +175,7 @@ error_code message_broker_impl::connect_lockfree(
 
    if ( !socket )
    {
-      LOG(error) << "failed to create socket";
+      LOG(error) << "Failed to create socket";
       disconnect_lockfree();
       return error_code::failure;
    }
@@ -184,7 +184,7 @@ error_code message_broker_impl::connect_lockfree(
 
    if ( err != AMQP_STATUS_OK )
    {
-      LOG(error) << "failed to open socket";
+      LOG(error) << "Failed to open socket";
       disconnect_lockfree();
       return error_code::failure;
    }
@@ -234,7 +234,7 @@ error_code message_broker_impl::connect( const std::string& url ) noexcept
 
    if( result != AMQP_STATUS_OK )
    {
-      LOG(error) << "unable to parse provided amqp url";
+      LOG(error) << "Unable to parse provided AMQP url";
       return error_code::failure;
    }
 

--- a/libraries/mq/request_handler.cpp
+++ b/libraries/mq/request_handler.cpp
@@ -259,12 +259,14 @@ void request_handler::publisher( std::shared_ptr< message_broker > publisher_bro
             if ( m->correlation_id )
                correlation_id = *m->correlation_id;
 
-            LOG(error) << "failed to ack correlation_id: " << correlation_id << ", delivery_tag: " << m->delivery_tag;
+            LOG(error) << "Failed to ack";
+            LOG(error) << " -> correlation_id: " << correlation_id;
+            LOG(error) << " -> delivery_tag:   " << m->delivery_tag;
          }
       }
       else
       {
-         LOG(error) << "an error has occurred while publishing message";
+         LOG(error) << "An error has occurred while publishing message";
       }
    }
 }
@@ -285,15 +287,24 @@ void request_handler::consumer( std::shared_ptr< message_broker > broker )
 
       if ( result.first != error_code::success )
       {
-         LOG(error) << "failed to consume message";
+         LOG(error) << "Failed to consume message";
          continue;
       }
 
       if ( !result.second )
       {
-         LOG(error) << "consumption succeeded but resulted in an empty message";
+         LOG(error) << "Consumption succeeded but resulted in an empty message";
          continue;
       }
+
+      LOG(debug) << "Received message";
+      LOG(debug) << " -> correlation_id: " << *result.second->correlation_id;
+      LOG(debug) << " -> exchange:       " << result.second->exchange;
+      LOG(debug) << " -> routing_key:    " << result.second->routing_key;
+      LOG(debug) << " -> content_type:   " << result.second->content_type;
+      LOG(debug) << " -> reply_to:       " << *result.second->reply_to;
+      LOG(debug) << " -> delivery_tag:   " << result.second->delivery_tag;
+      LOG(debug) << " -> data:           " << result.second->data;
 
       try
       {

--- a/libraries/mq/request_handler.cpp
+++ b/libraries/mq/request_handler.cpp
@@ -249,25 +249,10 @@ void request_handler::publisher( std::shared_ptr< message_broker > publisher_bro
 
       auto r = publisher_broker->publish( *m );
 
-//      if ( r == error_code::success )
-//      {
-//         r = consumer_broker->ack_message( m->delivery_tag );
-//         if ( r != error_code::success )
-//         {
-//            std::string correlation_id = "<none>";
-//
-//            if ( m->correlation_id )
-//               correlation_id = *m->correlation_id;
-//
-//            LOG(error) << "Failed to ack";
-//            LOG(error) << " -> correlation_id: " << correlation_id;
-//            LOG(error) << " -> delivery_tag:   " << m->delivery_tag;
-//         }
-//      }
-//      else
-//      {
-//         LOG(error) << "An error has occurred while publishing message";
-//      }
+      if ( r != error_code::success )
+      {
+         LOG(error) << "An error has occurred while publishing message";
+      }
    }
 }
 
@@ -298,11 +283,17 @@ void request_handler::consumer( std::shared_ptr< message_broker > broker )
       }
 
       LOG(debug) << "Received message";
-      LOG(debug) << " -> correlation_id: " << *result.second->correlation_id;
+
       LOG(debug) << " -> exchange:       " << result.second->exchange;
       LOG(debug) << " -> routing_key:    " << result.second->routing_key;
       LOG(debug) << " -> content_type:   " << result.second->content_type;
-      LOG(debug) << " -> reply_to:       " << *result.second->reply_to;
+
+      if ( result.second->correlation_id )
+         LOG(debug) << " -> correlation_id: " << *result.second->correlation_id;
+
+      if ( result.second->reply_to )
+         LOG(debug) << " -> reply_to:       " << *result.second->reply_to;
+
       LOG(debug) << " -> delivery_tag:   " << result.second->delivery_tag;
       LOG(debug) << " -> data:           " << result.second->data;
 

--- a/libraries/mq/request_handler.cpp
+++ b/libraries/mq/request_handler.cpp
@@ -251,7 +251,16 @@ void request_handler::publisher( std::shared_ptr< message_broker > publisher_bro
 
       if ( r == error_code::success )
       {
-         consumer_broker->ack_message( m->delivery_tag );
+         r = consumer_broker->ack_message( m->delivery_tag );
+         if ( r != error_code::success )
+         {
+            std::string correlation_id = "<none>";
+
+            if ( m->correlation_id )
+               correlation_id = *m->correlation_id;
+
+            LOG(error) << "failed to ack correlation_id: " << correlation_id << ", delivery_tag: " << m->delivery_tag;
+         }
       }
       else
       {

--- a/libraries/mq/request_handler.cpp
+++ b/libraries/mq/request_handler.cpp
@@ -173,7 +173,7 @@ error_code request_handler::add_msg_handler(
       if ( queue_res.first != error_code::success )
          return queue_res.first;
 
-      ec = _consumer_broker->bind_queue( queue_res.second, exchange, routing_key, !competing_consumer );
+      ec = _consumer_broker->bind_queue( queue_res.second, exchange, routing_key, true );
       if ( ec != error_code::success )
          return ec;
 
@@ -249,25 +249,25 @@ void request_handler::publisher( std::shared_ptr< message_broker > publisher_bro
 
       auto r = publisher_broker->publish( *m );
 
-      if ( r == error_code::success )
-      {
-         r = consumer_broker->ack_message( m->delivery_tag );
-         if ( r != error_code::success )
-         {
-            std::string correlation_id = "<none>";
-
-            if ( m->correlation_id )
-               correlation_id = *m->correlation_id;
-
-            LOG(error) << "Failed to ack";
-            LOG(error) << " -> correlation_id: " << correlation_id;
-            LOG(error) << " -> delivery_tag:   " << m->delivery_tag;
-         }
-      }
-      else
-      {
-         LOG(error) << "An error has occurred while publishing message";
-      }
+//      if ( r == error_code::success )
+//      {
+//         r = consumer_broker->ack_message( m->delivery_tag );
+//         if ( r != error_code::success )
+//         {
+//            std::string correlation_id = "<none>";
+//
+//            if ( m->correlation_id )
+//               correlation_id = *m->correlation_id;
+//
+//            LOG(error) << "Failed to ack";
+//            LOG(error) << " -> correlation_id: " << correlation_id;
+//            LOG(error) << " -> delivery_tag:   " << m->delivery_tag;
+//         }
+//      }
+//      else
+//      {
+//         LOG(error) << "An error has occurred while publishing message";
+//      }
    }
 }
 

